electrolytes/solvation_shell_utils.py

```diff
@@ -0,0 +1,168 @@
+    shells: List[Set[int]], st: Structure, solute_res_names: List[str]
+) -> List[Structure]:
+    """
+    Filter out shells that contain solute atoms.
+    Args:
+        shells: List of sets of atom indices that correspond to shells
+        st: Entire structure from the PDB file
+        solute_res_names: List of residue names that correspond to solute atoms in the simulation
+    Returns:
+        List of Schrodinger structures that do not contain solute atoms
+    """
+    actual_shells = [st.extract(shell, copy_props=True) for shell in shells]
+    filtered_shells = [
+        shell
+        for shell in actual_shells
+        if not any(
+            at.getResidue().pdbres.strip() in solute_res_names for at in shell.atom
+        )
+    ]
+    return filtered_shells
+
+
+def expand_shell(
+    st: Structure,
+    shell_ats: Set[int],
+    central_solute: int,
+    radius: float,
+    solute_res_names: List[str],
+    max_shell_size: int = 200,
+) -> Structure:
+    """
+    Expands a solvation shell. If there are any (non-central) solutes present in the shell,
+    recursively include shells around those solutes.
+    First, gets the molecule numbers of solute molecules that are within the radius
+    and not already expanded around. Then, continuously expand around them as long as we don't hit an atom limit.
+    Args:
+        st: Entire structure from the PDB file
+        shell_ats: Set of atom indices (of `st`) in a shell (1-indexed)
+        central_solute: Molecule index (of 'st') of the central solute in the shell
+        radius: Solvation radius (Angstroms) to consider
+        solute_res_names: List of residue names that correspond to solute atoms in the simulation
+        max_shell_size: Maximum size (in atoms) of the expanded shell
+    Returns:
+        Structure object containing the expanded solvation shell
+    """
+    # TODO: I'm not sure why we have to do this, but sometimes the central solute is not in the shell
+    shell_ats.add(central_solute)
+
+    solutes_included = set([central_solute])
+    while len(shell_ats) < max_shell_size:
+        old_shell_ats = shell_ats.copy()
+        new_solutes = set()
+        for at in shell_ats:
+            # If atom is part of a non-central solute molecule - should expand the shell
+            if (
+                st.atom[at].molecule_number not in solutes_included
+                and st.atom[at].getResidue().pdbres.strip() in solute_res_names
+            ):
+                new_solutes.add(st.atom[at].molecule_number)
+
+        if new_solutes:
+            # add entire residues within solvation shell radius of any extra solute atoms
+            shell_ats.update(
+                analyze.evaluate_asl(
+                    st,
+                    f'fillres within {radius} mol {",".join([str(i) for i in new_solutes])}',
+                )
+            )
+            solutes_included.update(new_solutes)
+        else:
+            break
+
+    shell_ats = old_shell_ats
```

Comment:
```suggestion
    while True:
        new_solutes = set()
        for at in shell_ats:
            # If atom is part of a non-central solute molecule - should expand the shell
            if (
                st.atom[at].molecule_number not in solutes_included
                and st.atom[at].getResidue().pdbres.strip() in solute_res_names
            ):
                new_solutes.add(st.atom[at].molecule_number)

        if new_solutes:
            # add entire residues within solvation shell radius of any extra solute atoms
            new_shell_ats = shell_ats.union(
                analyze.evaluate_asl(
                    st,
                    f'fillres within {radius} mol {",".join([str(i) for i in new_solutes])}',
                )
            )
            if new_shell_ats <= max_shell_size:
                shell_ats = new_shell_ats
                solutes_included.update(new_solutes)
            else:
                break
        else:
            break
```

My suggestion just to be a little more explicit. Or (I think better yet):

```suggestion
    def get_new_solutes(st, shell_ats, solutes_included, solute_res_names):
        new_solutes = set()
        for at in shell_ats:
            # If atom is part of a non-central solute molecule - should expand the shell
            if (
                st.atom[at].molecule_number not in solutes_included
                and st.atom[at].getResidue().pdbres.strip() in solute_res_names
            ):
                new_solutes.add(st.atom[at].molecule_number)
        return new_solutes
 
    new_solutes = get_new_solutes(st, shell_ats, solutes_included, solute_res_names)
    while new_solutes:
        # add entire residues within solvation shell radius of any extra solute atoms
        new_shell_ats = shell_ats.union(
            analyze.evaluate_asl(
                st,
                f'fillres within {radius} mol {",".join([str(i) for i in new_solutes])}',
            )
        )
        if new_shell_ats <= max_shell_size:
            shell_ats = new_shell_ats
            solutes_included.update(new_solutes)
            new_solutes = get_new_solutes(st, shell_ats, solutes_included, solute_res_names)
        else:
            break
```

Reply:
Changed.